runtime/extern.c

```diff
@@ -906,6 +1005,25 @@ static intnat extern_value(struct caml_extern_state* s, value v, value flags,
   extern_rec(s, v);
   /* Record end of output */
   close_extern_output(s);
+  /* Compress if requested */
+#ifdef HAS_ZSTD
+  if (s->extern_flags & COMPRESSED) {
+    uintnat uncompressed_len = extern_output_length(s);
+    extern_compress_output(s);
+    res_len = extern_output_length(s);
+    /* Write the header in compressed format */
+    store32(header, Intext_magic_number_compressed);
+    int pos = 5, len;
+    len = storevlq(header + pos, res_len); pos += len;
+    len = storevlq(header + pos, uncompressed_len); pos += len;
+    len = storevlq(header + pos, s->obj_counter); pos += len;
+    len = storevlq(header + pos, s->size_32); pos += len;
```

Comment:
This seems like a good place to do a generic COMPAT_32 check for overflow given that on the read path there might be a failure on 32-bit platforms. See also the COMPAT_32 check that would've been done in the Intext_magic_number_big mode that seems to be missing here.

Reply:
Nice catch!  I completely forgot about the interaction of `Compression` with `Compat_32`...  Missing checks added in 853b6500f4 .